pkg/solver/singlehost.go

```diff
@@ -0,0 +1,325 @@
+
+			if len(s.Labels[k]) == 0 {
+				s.Labels[k] = v
+			}
+		}
+
+		if s.Annotations == nil {
+			s.Annotations = map[string]string{}
+		}
+
+		for k, v := range annos {
+
+			if len(s.Annotations[k]) == 0 {
+				s.Annotations[k] = v
+			}
+		}
+	}
+
+	// k, now we have to create our own objects for configuring the gateway
+	configMaps := c.getGatewayConfigMaps(cheManager, workspaceMeta.WorkspaceId, routing)
+
+	syncer := sync.New(c.client, c.scheme)
+
+	for _, cm := range configMaps {
+		_, _, err := syncer.Sync(context.TODO(), nil, &cm, configMapDiffOpts)
+		if err != nil {
+			return solvers.RoutingObjects{}, err
+		}
+
+	}
+	return objs, nil
+}
+
+func (c *CheRoutingSolver) singlehostExposedEndpoints(manager *dwoche.CheManager, workspaceID string, endpoints map[string]dwo.EndpointList, routingObj solvers.RoutingObjects) (exposedEndpoints map[string]dwo.ExposedEndpointList, ready bool, err error) {
+	if manager.Status.GatewayPhase != dwoche.GatewayPhaseEstablished {
+		return nil, false, nil
+	}
+
+	host := manager.Status.GatewayHost
+	if host == "" {
+		return nil, false, nil
+	}
+
+	exposed := map[string]dwo.ExposedEndpointList{}
+
+	for machineName, endpoints := range endpoints {
+		exposedEndpoints := dwo.ExposedEndpointList{}
+		for _, endpoint := range endpoints {
+			if endpoint.Exposure != dw.PublicEndpointExposure {
+				continue
+			}
+
+			var scheme string
+			if endpoint.Protocol == "" {
+				scheme = "http"
+			} else {
+				scheme = string(endpoint.Protocol)
+			}
+
+			if scheme != "http" && scheme != "https" {
+				// we cannot expose non-http endpoints publicly, because ingresses/routes only support http(s)
+				continue
+			}
+
+			if endpoint.Secure {
+				scheme = "https"
+
+				// TODO this should also do the magic of ensuring user authentication however we are going to do it
+				// in the future
+			}
+
+			publicURLPrefix := getPublicURLPrefixForEndpoint(workspaceID, machineName, endpoint)
+
+			publicURL := scheme + "://" + ensureDoesntEndWithSlash(host) + ensureDoesntEndWithSlash(publicURLPrefix) + ensureStartsWithSlash(endpoint.Path)
+
+			attrs := map[string]string{}
+			err := endpoint.Attributes.Into(&attrs)
+			if err != nil {
+				return nil, false, err
+			}
+
+			exposedEndpoints = append(exposedEndpoints, dwo.ExposedEndpoint{
+				Name:       endpoint.Name,
+				Url:        publicURL,
+				Attributes: attrs,
+			})
+		}
+		exposed[machineName] = exposedEndpoints
+	}
+
+	return exposed, true, nil
+}
+
+func (c *CheRoutingSolver) getGatewayConfigMaps(cheManager *dwoche.CheManager, workspaceID string, routing *dwo.WorkspaceRouting) []corev1.ConfigMap {
+	restrictedAnno, setRestrictedAnno := routing.Annotations[config.WorkspaceRestrictedAccessAnnotation]
+
+	labels := defaults.GetLabelsForComponent(cheManager, "gateway-config")
+	labels[config.WorkspaceIDLabel] = workspaceID
+	if setRestrictedAnno {
+		labels[config.WorkspaceRestrictedAccessAnnotation] = restrictedAnno
+	}
```

Comment:
I was trying to do the same thing that is being done in the `workspacerouting_controller.go` to services/ingresses/routes without actually understanding if this has any consequence since this is applied to a configmap in a different namespace from the namespace of the workspace.

Reply:
For configmaps it'll have no effect, since our webhooks don't listen for those (yet...). The main use cases are [don't allow pods/exec](https://github.com/devfile/devworkspace-operator/blob/c4ad8ce4d2ab25346daecab9ea1db3e933cbe081/webhook/workspace/handler/exec.go#L26-L53) and [don't allow objects to be modified](https://github.com/devfile/devworkspace-operator/blob/master/webhook/workspace/handler/immutable.go). Only the creator and the DevWorkspaceOperator can modify objects with `restricted-access` set.

Reply:
Ok, I'll leave the annotation in then, because basically the config maps are "restricted-access". Any change to them will be reverted by the operator to conform to what's in the routing.